assets/pages/regular-contributions/components/formFields.jsx

```diff
@@ -127,27 +141,46 @@ function countriesDropdown(
 
 function NameForm(props: PropTypes) {
 
+  const showFirstNameError = emptyInputField(props.firstName.value) && props.firstName.shouldValidate;
```

Comment:
As discussed...

My hope with using the Constraint Validation API was to have a uniform interface for validation (both the declarative constraints in the HTML, and the `element.validity.valid` for checking the constraint are met). So it seems a shame to not be able to use `element.validity.valid` (and thereby the HTML fields like `required`) in individual element validation like we do in the form validation function above.

One option might be to use [refs](https://reactjs.org/docs/refs-and-the-dom.html) to get access to the DOM elements that have the `validity.valid` API available on them. I'm not sure what @Ap0c might think of that and whether it breaks our React philosophy (though since we're only reading from the DOM API maybe there's a case for it).

The other option - which we decided might be simpler - was to implement our own uniform interface in React-land by requiring an `isValid` function on the form elements. The downside there would be that it wouldn't make use of the declarative style with attributes like `required` - the fact that an element is required or needs to confirm to a regex would be defined inside the `isValid` function. Maybe this is fine.

Reply:
My 2 cents: everything the browser does should be leveraged. That's a whole class of bugs easily avoided and maintenance time spared.

Reply:
(so yes, refs)

Reply:
I'm inclined to agree!